src/LongRange/tests/test_temp.cpp

```diff
@@ -0,0 +1,104 @@
+//  double norm;
+//  inline double operator()(double r, double rinv) {return rinv;}
+//  void reset(ParticleSet& ref) {norm=4.0*M_PI/ref.LRBox.Volume;}
+//  inline double Xk(double k, double rc) {return -norm/(k*k)*std::cos(k*rc);}
+//  inline double Fk(double k, double rc) {return -Xk(k, rc);}
+//  inline double integrate_r2(double r) const {return 0.5*r*r;}
+//  inline double df(double r) {return 0;} // ignore derivatives for now
+//  void reset(ParticleSet& ref, double rs) {reset(ref);} // ignore rs
+//};
+template<class T = double>
+struct EslerCoulomb3D
+{ // copied from LRCoulombSingleton::CoulombFunctor
+  T NormFactor;
+  inline EslerCoulomb3D() {}
+  void reset(ParticleSet& ref) { NormFactor = 4.0 * M_PI / ref.LRBox.Volume; }
+  void reset(ParticleSet& ref, T rs) { NormFactor = 4.0 * M_PI / ref.LRBox.Volume; }
+  inline T operator()(T r, T rinv) { return rinv; }
+  inline T df(T r) { return -1.0 / (r * r); }
+  inline T df2(T r) { return 2.0 / (r * r * r); }
+  inline T Vk(T k) { return NormFactor / (k * k); }
+
+  inline T Xk_dk(T k) { return 0.0; }
+  inline T Fk(T k, T rc) { return NormFactor / (k * k) * std::cos(k * rc); }
+  inline T Xk(T k, T rc) { return -NormFactor / (k * k) * std::cos(k * rc); }
+
+  inline T dVk_dk(T k) { return -2 * NormFactor / k / k / k; }
+  inline T dFk_dk(T k, T rc) { return -NormFactor / k / k * (2.0 / k * std::cos(k * rc) + rc * std::sin(k * rc)); }
+
+  inline T dXk_dk(T k, T rc) { return NormFactor / k / k * (2.0 / k * std::cos(k * rc) + rc * std::sin(k * rc)); }
+
+  inline T integrate_r2(T r) const { return 0.5 * r * r; }
+};
+
+/** evalaute bare Coulomb in 3D using LRHandlerTemp
+ */
+TEST_CASE("temp3d", "[lrhandler]")
+{
+  CrystalLattice<OHMMS_PRECISION, OHMMS_DIM> Lattice;
+  Lattice.BoxBConds = true;
+  Lattice.LR_dim_cutoff = 30.;
+  Lattice.R.diagonal(5.0);
+  Lattice.reset();
+  REQUIRE(Lattice.Volume == Approx(125));
+  Lattice.SetLRCutoffs(Lattice.Rv);
+  //Lattice.printCutoffs(app_log());
+  REQUIRE(Lattice.LR_rc == Approx(2.5));
+  REQUIRE(Lattice.LR_kc == Approx(12));
+
+  ParticleSet ref; // handler needs ref.SK.KLists
+  ref.Lattice = Lattice;  // !!!! crucial for access to Volume
+  ref.LRBox = Lattice;  // !!!! crucial for S(k) update
+  StructFact *SK = new StructFact(ref, Lattice.LR_kc);
+  ref.SK = SK;
+  LRHandlerTemp<EslerCoulomb3D<double>, LPQHIBasis> handler(ref);
+
+  handler.initBreakup(ref);
+  REQUIRE(handler.MaxKshell == 78);
+  REQUIRE(handler.LR_rc == Approx(2.5));
+  REQUIRE(handler.LR_kc == -1); // ?? YY: shouldn't the internal be updated?
+
+  mRealType r, dr, rinv;
+  mRealType vsr, vlr;
+  int nr = 101;
+  dr = 5.0/nr;  // L/[# of grid points]
+  for (int ir=1;ir<nr;ir++)
+  {
+    r = ir*dr;
+    rinv = 1./r;
+    vsr = handler.evaluate(r, rinv);
+    vlr = handler.evaluateLR(r);
+    //// short-range part must vanish after rcut
+    //if (r>2.5) REQUIRE(vsr == Approx(0.0));
+    //// ?? YY: may it not?
```

Comment:
Uh.  I would think vsr should be zero here.  It's not??

Reply:
I will think about this.

Reply:
OK.  My knee jerk reaction is that LRHandlerTemp is playing fast and loose with the evaluation of short-stuff for r>r_c.  v_{lr}(r) = \sum c_i h_i(r) inside r_c where c_i is a coefficient, h_i(r) is a smooth basis function with compact support.  However, v_{lr}(r)=v(r) at r>=r_c (and the matching at r=r_c should be continuous).  This implies that the short-range piece, given by v_{sr}=v(r)-v_{lr}(r) is v(r)-\sum c_i h_i(r) for r<r_c (which is what the code does), and strictly 0 for r>r_c.  The fact that it does not return 0 in this case is an actual bug.  That this has not manifested in the mainline QMCPACK is because in PBC's, CoulombPBCAA and CoulombPBCAB spline the short-range piece only up to r<=r_c and evaluate to 0 for r>r_c.  This should definitely be fixed, because if I chose not to spline, I would get the wrong answers.

Reply:
If you look at lrDf(), all real-space functions should have that logic.

Reply:
@rcclay are you saying the `LRHanlderTemp::evaluate` should have an if statement that sets it to zero when r>rc?

This effectively makes the short-range potential a piece-wise potential with a jump at L/2. Would this not cause any problems?